engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,843 @@
+	currTable := g.normalTables[sourceTable]
+	defer func() {
+		g.normalTables[sourceTable] = currTable
+	}()
+
+	g.normalTables[sourceTable] = prevTableStmt
+	oldJoined, err := g.joinTables(normal)
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	postTable := genCmpTable(postTableStmt)
+	g.normalTables[sourceTable] = postTableStmt
+	newJoined, err := g.joinTables(normal)
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	cmp, err := oldJoined.Compare(newJoined)
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	if cmp <= 0 {
+		if col, err2 := optimism.AddDifferentFieldLenColumns("", ddl, oldJoined, newJoined); err2 != nil {
+			// check for add column with a larger field len
+			return "", err2
+		} else if _, err2 = optimism.AddDifferentFieldLenColumns("", ddl, postTable, newJoined); err2 != nil {
+			// check for add column with a smaller field len
+			return "", err2
+		} else if len(col) > 0 && (g.droppedColumnsStore.HasDroppedColumn(ctx, col, sourceTable) || slices.Contains(newDroppedColumns, col)) {
+			return "", errors.Errorf("add column %s that wasn't fully dropped in downstream. ddl: %s", col, ddl)
+		}
+	}
+
+	if cmp >= 0 {
+		if col, err2 := optimism.GetColumnName("", ddl, ast.AlterTableDropColumn); err2 != nil {
+			return "", err2
+		} else if len(col) > 0 {
+			return col, nil
+		}
+	}
+	return "", nil
+}
+
+func (g *shardGroup) gcDroppedColumns(ctx context.Context) error {
+	state, err := g.droppedColumnsStore.Get(ctx)
+	if err != nil {
+		if errors.Cause(err) == metadata.ErrStateNotFound {
+			return nil
+		}
+		return err
+	}
+
+	droppedColumns := state.(*metadata.DroppedColumns)
+	cacheStmts := make(map[metadata.SourceTable]string)
+OutLoop:
+	for col := range droppedColumns.Cols {
+		for _, tbStmt := range g.normalTables {
+			if tbStmt == "" {
+				continue
+			}
+			if cols := getColumnNames(tbStmt); slices.Contains(cols, col) {
+				continue OutLoop
+			}
+		}
+
+		for sourceTable := range g.normalTables {
+			tbStmt, ok := cacheStmts[sourceTable]
+			if !ok {
+				tbStmt, err = g.tableAgent.FetchTableStmt(ctx, g.id, g.cfg, sourceTable)
+				if err != nil {
+					if strings.HasPrefix(err.Error(), "table info not found") {
+						continue
+					}
+					return err
+				}
+			}
+
+			if cols := getColumnNames(tbStmt); slices.Contains(cols, col) {
+				continue OutLoop
+			}
+		}
+		if err := g.droppedColumnsStore.DelDroppedColumn(ctx, col); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// isResoved means all tables in the group are resolved.
+// 1. no conflict ddls waiting
+// 2. all dropped column has done
+// 3. all shard tables stmts are same.
+func (g *shardGroup) isResolved(ctx context.Context) bool {
+	if len(g.conflictTables) != 0 {
+		return false
+	}
+	if _, err := g.droppedColumnsStore.Get(ctx); err == nil || errors.Cause(err) != metadata.ErrStateNotFound {
```

Comment:
```suggestion
	if _, err := g.droppedColumnsStore.Get(ctx); errors.Cause(err) != metadata.ErrStateNotFound {
```

seems `errors.Cause` is clear enough